### DISCUSSION TITLE
Does the Provider absolutely always have to be static and final?

### DISCUSSION THREAD

**Discussion by aoisensi** [rrousselGit/riverpod#1679] [Q&A] [answered]
I am developing a complex application, is such an implementation allowed?
like
```dart
final cachesProvider =
    Provider.family<Cache, String>(((ref, arg) => Cache(arg)));

class Cache {
  Cache(this.userId);

  final String userId;
  final _usersProvider = StateProvider<Map<String, User>>((ref) => {});
  late final usersProvider = StateProvider.family<User, String>((ref, arg) {
    return ref.watch(_usersProvider)[arg]!;
  });
  //
  //  MORE CACHE PROVIDER ...
  //
}
```
I just want to construct Provider dynamicaly by other Provider.
Is this ok from Riverpod perspective?

**Comment by rrousselGit** [comment, marked as answer]
No.

Providers should always be global finals.

**Reply by aoisensi** [reply]
Thank you for your answer.
So, what if I want to have multiple Providers provided by a library, for example?
Is there any good design pattern?

**Reply by rrousselGit** [reply]
I'd need a more specific question. You can have as many providers as you want, but I don't really understand your question.

**Reply by aoisensi** [reply]
Sorry for bad English.
I am building a client app and a API library for Discord.
The app allows multiple accounts to be logged in at the same time.
Since Discord's API response resources change frequently, I am thinking of using Riverpod to cache them on the library side.
However, the library will mostly be used for login purposes for a single account.
I don't know how to design the Provider of the library in these cases.
Thank you for reading.

**Reply by TimWhiting** [reply]
My approach to solving a similar problem is this:

```dart
final accountIdProvider = Provider((ref) => throw UnimplementedError());

final discordResourceProvider = Provider((ref) => ...., dependencies: [accountIdProvider]);
....

// in widget tree
ProviderScope(overrides: [accountIdProvider.overrideWithValue(accountId)], child: ...YourAppViewForThisAccount));
```

This makes sure that each account gets a separate resource providers, but they all can be in the tree / keep their state simultaneously, and I don't have to pass the accountId everywhere in my app like I would have to if I were to use families.

In other words, if you want to create a set of providers that can be duplicated or configured, you just have to set up a configuration or id provider that the user of the providers can override, and make sure all of your providers that should be duplicated (everything except the configuration / id provider) list the configuration / id provider in their dependencies (this causes them to be duplicated for that particular subtree).

This is how I test multiplayer games without having to run 10+ separate instances of an app, which would be a pain.

**Reply by aoisensi** [reply]
hmmm Thank you.
